Add support for 4K response data

xyan264 opened the pull request:
New function `switchtec_nvme_admin_passthru` from switchtec-user library supports 4K response data from SRIOV devices

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
+	uint8_t nvme_data[4096];
+};
+
+int pax_nvme_submit_admin_passthru_v1(int fd, struct nvme_passthru_cmd *cmd)
```
_v1 seems not clear, how about _1k_rsp?

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+	ret = switchtec_nvme_admin_passthru(pax->dev, pax->pdfid,
+					    data_len, &req,
+					    &rsp_len, &rsp);
+	if (ret && errno == EBADMSG) {
```
EBADMSG from firmware? Should be UNSUPPORTED?

xyan264 replied:
Yes this is a problem. Let's discuss.

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
 	return status;
 }
 
+int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
```
We should have another _4k_rsp function for the new implementation, and this function just try _4k_rsp and 1k_rsp and return.

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
 };
 
+struct switchtec_adm_passthru_nvme_req
+{
```
'{' should be in the same line of struct

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -48,8 +49,20 @@ struct switchtec_device_manage_nvme_rsp
 
-int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+struct switchtec_adm_passthru_nvme_rsp
+{
```
ditto

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -100,6 +113,61 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+					    &rsp_len, &rsp);
+	if (ret && errno == EBADMSG) {
+		/* if device does not support new function, */
+		/* call legacy function instead */
```
Normally we do multi-line comment like this 
```
/*
 * This is an example of multi-line comment. we'll wrap when we reach
 * 80 characters.
 */
```

kelvin-cao commented on plugins/microchip/switchtec-nvme-device.c:
```diff
@@ -100,6 +109,70 @@ int pax_nvme_submit_admin_passthru(int fd, struct nvme_passthru_cmd *cmd)
+		ret = pax_nvme_submit_admin_passthru_1k_rsp(fd, cmd);
+
+	if (ret)
+		switchtec_perror("pax_nvme_submit_admin_passthru");
```
We should not do the error message output in the API, instead we leave it to the callers.

xyan264 commented:
@kelvin-cao Code updated with the following changes:
- removed error prints from `pax_nvme_submit_admin_passthru`
- added `errno` checking:
```	ret = pax_nvme_submit_admin_passthru_4k_rsp(fd, cmd);
	if (ret && errno == INVALID_MRPC_CMD)
		ret = pax_nvme_submit_admin_passthru_1k_rsp(fd, cmd);
```
